src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,223 @@
+{
+	public ExecOpRequestTPFatTPFServerWithTranslation( final TriplePatternRequest req, final TPFServer fm ) {
+		super( req, fm );
+
+		assert fm.getVocabularyMapping() != null;
+	}
+
+	
+	@Override
+	protected void _execute(IntermediateResultElementSink sink, ExecutionContext execCxt)
+			throws ExecOpExecutionException {
+		
+		FederationAccessManager fedAccessMgr = execCxt.getFederationAccessMgr();
+		final SPARQLGraphPattern reqTranslation = fm.getVocabularyMapping().translateTriplePattern(req.getQueryPattern());
+		Iterator<SolutionMapping> res;
+		if(reqTranslation instanceof TriplePattern) {
+			res = handleTriplePattern((TriplePattern) reqTranslation, fedAccessMgr).iterator();
+		}
+		else if(reqTranslation instanceof SPARQLUnionPattern) {
+			res = handleUnionPattern(((SPARQLUnionPattern) reqTranslation), fedAccessMgr);
+		}
+		else if(reqTranslation instanceof SPARQLGroupPattern) {
+			res = handleGroupPattern(((SPARQLGroupPattern) reqTranslation), fedAccessMgr);
+		}
+		else if(reqTranslation instanceof BGP) {
+			res = handleBGP((BGP) reqTranslation, fedAccessMgr);
+		}
+		else {
+			throw new ExecOpExecutionException(reqTranslation.toString(), this);
+		}
+		
+		while(res.hasNext()) {
+			sink.send(res.next());
+		}
+		
+	}
+	
+	protected Iterable<SolutionMapping> handleTriplePattern(final TriplePattern tp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException{
+		
+		final TriplePatternRequest newReq = new TriplePatternRequestImpl(tp);
+		final ExecOpRequestTPFatTPFServer op = new ExecOpRequestTPFatTPFServer(newReq, fm);
+		final MaterializingIntermediateResultElementSinkWithTranslation sink = new MaterializingIntermediateResultElementSinkWithTranslation(fm.getVocabularyMapping());
+		
+		TPFResponse currentPage = null;
+		while ( currentPage == null || ! op.isLastPage(currentPage) ) {
+			// create the request for the next page (which is the first page if currentPage is null)
+			final TPFRequest pageRequest = op.createPageRequest(currentPage);
+
+			// perform the page request
+			try {
+				currentPage = op.performPageRequest( pageRequest, fedAccessMgr );
+			}
+			catch ( final FederationAccessException e ) {
+				throw new ExecOpExecutionException("Issuing a page request caused an exception: " + e.toString(), this);
+			}
+			
+			// consume the matching triples retrieved via the page request
+			op.consumeMatchingTriples( currentPage.getPayload(), sink );
+
+		}	
+	
+		return sink.getMaterializedIntermediateResult();
+
+	}
+	
+	protected Iterator<SolutionMapping> handleUnionPattern(final SPARQLUnionPattern up, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SolutionMapping> unionTranslation = null;
+		for(final SPARQLGraphPattern i : up.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				if (unionTranslation == null) {
+					unionTranslation = handleTriplePattern(((TriplePattern) i), fedAccessMgr).iterator();
+				} else {
```

Comment:
I think that it would be best to iterate through the iterator of `up.getSubPatterns()`. That way, you can consume the first element first, and store it in `unionTranslation` and then iterate through the rest to create the union tree.

Furthermore, you can add a generic `handlePattern` method that evaluates all the `instanceof` conditions and calls the `handle...` method appropriate for the pattern type, which will simplify all the code of this class.